plugins/xlsxreader.cpp

```diff
@@ -0,0 +1,181 @@
+/*
+ * Wrapper for xlsxio_read library functions.
+ *
+ * Sample usage:
+ *
+ *  std::string filename = "sample_file.xlsx";
+ *  xlsxioreader xlsxfile = XlsxReader::open_xlsx_file(filename);
+ *  if (xlsxfile == NULL) {
+ *      printf("cannot open file: '%s'", filename.c_str());
+ *      return false;
+ *  }
+ *  auto sheetNames = XlsxReader::list_sheets(xlsxfile);
+ *  for (auto sheetName = sheetNames.begin();
+ *       sheetName != sheetNames.end();
+ *       ++sheetName) {
+ *      printf("reading sheet: %s\n", sheetName->c_str());
+ *      xlsxioreadersheet xlsxsheet =
+ *              XlsxReader::open_sheet(xlsxfile, *sheetName);
+ *      if (xlsxsheet == NULL) {
+ *          printf("cannot open sheet: '%s'", sheetName->c_str());
+ *          continue;
+ *      }
+ *      std::string value;
+ *      int row_num = 1;
+ *      while (XlsxReader::GetNextRow(xlsxsheet)) {
+ *          std::string s;
+ *          printf("%d:\t", row_num);
+ *          while (XlsxReader::GetNextCell(xlsxsheet, s)) {
+ *              printf("%s\t", s.c_str());
+ *          }
+ *          printf("\n");
+ *          ++row_num;
+ *      }
+ *      XlsxReader::close_sheet(xlsxsheet);
+ *  }
+ *  XlsxReader::close_xlsx_file(xlsxfile);
+ *  return true;
+ */
+
+#define BUILD_XLSXIO_STATIC
+#include <xlsxio_read.h>
+
+#include "DataFuncs.h"
+#include "LuaTools.h"
+#include "PluginManager.h"
+
+using namespace DFHack;
+
+DFHACK_PLUGIN("xlsxreader");
+
+struct handle_identity : public compound_identity {
+    handle_identity(): compound_identity(0, nullptr, nullptr, "handle") {};
+    DFHack::identity_type type() override { return IDTYPE_OPAQUE; }
+};
+
+struct xlsx_file_handle {
+    const xlsxioreader handle;
+    xlsx_file_handle(xlsxioreader handle): handle(handle) {}
+    static handle_identity _identity;
+};
+
+struct xlsx_sheet_handle {
+    const xlsxioreadersheet handle;
+    xlsx_sheet_handle(xlsxioreadersheet handle): handle(handle) {}
+    static handle_identity _identity;
+};
+
+handle_identity xlsx_file_handle::_identity;
+handle_identity xlsx_sheet_handle::_identity;
```

Comment:
It might be a good idea to give these separate names, i.e. changing the parameter that's hardcoded to `"handle"` above. As far as I know, the names are still human-facing, but it could be confusing to have two different types, both named "handle", with functions that only accept one of them.

Out of curiosity, have you tried unloading this plugin while Lua code has access to a handle? If that crashes, I wouldn't be too surprised, but there are probably other ways to cause a crash when unloading some plugins, so it's probably an acceptable risk for now.

I haven't yet figured out if simply creating an object with these identities would be enough to cause a crash later if the plugin is unloaded, either when garbage collection kicks in or if some internal code collects a list of identities as they're registered and tries to access them later. If you haven't encountered a crash after some time in-game after running quickfort, and maybe after calling `collectgarbage()` from Lua, I suspect it's fine. I'd like to do a bit more research into this, though.

Reply:
I tested reloading the plugin while a handle is open, and everything worked ok:
DFHack# lua
[lua]# xlsxreader = require('plugins.xlsxreader')
[lua]# xlsx_file = xlsxreader.open_xlsx_file('blueprints/industry.xlsx')
[lua]# xlsx_sheet = xlsxreader.open_sheet(xlsx_file, 'dig')
[lua]# printall(xlsxreader.get_row(xlsx_sheet))
[expected output]
[lua]# printall(xlsxreader.get_row(xlsx_sheet))
[expected output]

then from a separate window:
$ ./dfhack-run reload xlsxreader

then back in DFHack:
[lua]# collectgarbage()
[lua]# printall(xlsxreader.get_row(xlsx_sheet))
[expected output]
[lua]# printall(xlsxreader.get_row(xlsx_sheet))
[expected output]
[lua]# xlsxreader.close_sheet(xlsx_sheet)
[lua]# xlsxreader.close_xlsx_file(xlsx_file)

I think it's expected that it works since xlsxreader holds all the state in the allocated handle. it doesn't have any global state to lose in a reload. the handles are just malloc'd, so it's not like the memory references become bad -- glibc is still the same

Reply:
The tricky part is that the addresses of the identities (like `xlsx_file_handle::_identity`) can change across reloads. I think those are used at least somewhere in the Lua wrapper layer. Your investigation sounds better than I had hoped for, though!

Reply:
ah, that's a very good point. Should I move the statics into the core library?

Reply:
library/include/PluginStatics.h and library/PluginStatics.cpp?

Reply:
I usually prefer to keep plugin things out of non-plugins, but I think this can be an exception (at least for now, until there's a better way to handle plugin-defined types).